File: terminale/architectures/se/ordonnancement-implementation/scripts/file.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
@Author: Christophe Viroulaud
@Time:   Lundi 18 Octobre 2021 23:11
"""


class Element():
    def __init__(self, d: int, s: object):
        self.donnees = d
        self.successeur = s


class File():
    def __init__(self):
        self.premier = None
        self.dernier = None

    def est_vide(self) -> bool:
        return self.premier == None

    def enfiler(self, e: int) -> None:
        nouveau = Element(e, None)

        if self.est_vide():
            # 1 seul élément: le premier est le dernier
            self.premier = nouveau
        else:
            # le dernier devient avant-dernier
            self.dernier.successeur = nouveau

        # le nouveau devient dernier
        self.dernier = nouveau

    def defiler(self) -> int:
        if not self.est_vide():
            res = self.premier.donnees
            self.premier = self.premier.successeur
            return res

    def __str__(self):
        c = self.premier
        s = ""
        while not c is None:
            s = s + str(c.donnees)+"|"
            c = c.successeur
        return "\u2BA4|" + s[:] + "\u2BA0"


if __name__ == "__main__":
    from random import randint

    a = File()
    for i in range(6):
        a.enfiler(randint(1, 20))
        print(a)

    for i in range(6):
        a.defiler()
        print(a)

File: terminale/architectures/se/ordonnancement-implementation/scripts/c/file.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Maillon Maillon;
struct Maillon {
  int val;
  Maillon *suiv;
};

typedef struct File File;
struct File {
  Maillon *premier;
  Maillon *dernier;
};

void enfiler(File **f, int v) {
  Maillon *nouv = malloc(sizeof(Maillon));
  nouv->val = v;
  nouv->suiv = NULL;
  if ((*f)->premier == NULL) {
    (*f)->premier = nouv;
  } else {
    (*f)->dernier->suiv = nouv;
  }
  (*f)->dernier = nouv;
}

int defiler(File **f) {
  if ((*f)->premier == NULL) {
    return -1;
  } else {
    int res = (*f)->premier->val;
    (*f)->premier = (*f)->premier->suiv;
    return res;
  }
}

int main(int argc, char *argv[]) {
  File *f = malloc(sizeof(File));
  f->premier = NULL;
  f->dernier = NULL;
  enfiler(&f, 5);
  enfiler(&f, 6);
  printf("%d", defiler(&f));
  return 0;
}